### ISSUE TITLE
`service_instances_name_index` needs rebuilding after 1.136.0 to 1.140.0 upgrade when using postgres

### ISSUE THREAD

**Issue by peterellisjones** [cloudfoundry/cloud_controller_ng#3081]
## Issue

Hi CAPI team,

We found that after upgrading from `1.136.0` to `1.140.0`, the index `service_instances_name_index` needed rebuilding via `REINDEX INDEX service_instances_name_index` in postgres.

## Context

After the upgrade, one of our users complained that they were unable to see their service using the `cf service $NAME` CF CLI command. We confirmed that a service that we were able to list via `cf services` did not appear when we tried to show the service via `cf service $NAME`, and instead the CF CLI gave the following error:

```
Showing info of service elk-logger in org  $ORG / space $SPACE as $USER
Service instance 'elk-logger' not found
```

We found that the DB queries used to build the CF API responses for `cf services` and `cf service $NAME` were using different indexes in postgres and that the index `service_instances_name_index` which the latter command used, was incomplete. For example, running `SELECT * FROM service_instances WHERE name = 'elk-logger'` returned far fewer results than expected (we expected > 100 rows, but got 3). Rebuilding the index via `REINDEX INDEX service_instances_name_index` fixed the issue.

We saw the same error on two separate environments on which we performed the same upgrade at different times. Both were fixed by rebuilding the `service_instances_name_index` index. The fact that the error was reproduced on two environments suggests a disk io error is unlikely as the cause and we did not see any errors in the postgres logs to indicate a disk problem.

We are not currently aware of other indexes that need rebuilding.

Although we saw the issue with a particular logging service which happened to be called `elk-logger`, we did not do a thorough check of other services before implementing the fix and therefore cannot say whether this issue affected just services with this name, all logging services, or all services generally.

We are using postgres 11.10 deployed via the bosh postgres release v43 https://bosh.io/releases/github.com/cloudfoundry/postgres-release?version=43, deployed on vSphere using the bosh-vsphere-esxi-ubuntu-jammy-go_agent stemcell v1.44. The Cloud Controller VMs are also using this stemcell.

cheers,

Pete

**Comment by peterellisjones** [comment]
Further investigation indicates this is likely an issue with postgres 11.9 that was fixed in 11.14. https://www.postgresql.org/docs/11/release-11-14.html

> Fix CREATE INDEX CONCURRENTLY to wait for the latest prepared transactions (Andrey Borodin)
> 
> Rows inserted by just-prepared transactions might be omitted from the new index, causing queries relying on the index to miss such rows. The previous fix for this type of problem failed to account for PREPARE TRANSACTION commands that were still in progress when CREATE INDEX CONCURRENTLY checked for them. As before, in installations that have enabled prepared transactions (max_prepared_transactions > 0), it's recommended to reindex any concurrently-built indexes in case this problem occurred when they were built.
> 
> Avoid race condition that can cause backends to fail to add entries for new rows to an index being built concurrently (Noah Misch, Andrey Borodin)
> 
> While it's apparently rare in the field, this case could potentially affect any index built or reindexed with the CONCURRENTLY option. It is recommended to reindex any such indexes to make sure they are correct.

I will update / close this issue once confirmed

**Comment by philippthun** [comment]
As we did not receive any further complains about this problem, I'm going to close the issue.
